lib/TPP/MapToBatchReduceGEMM.cpp

```diff
@@ -58,6 +83,45 @@ static bool isOutputOperand(linalg::LinalgOp linalgOp, OpOperand &operand) {
   return !isInputOperand(linalgOp, operand);
 }
 
+static LogicalResult checkVNNIAccessPatterns(linalg::LinalgOp linalgOp) {
+  SmallVector<AffineMap> maps;
+  for (OpOperand &operand : linalgOp->getOpOperands()) {
+    AffineMap map = linalgOp.getMatchingIndexingMap(&operand);
+    if (isInputOperand(linalgOp, operand)) {
+      if (map.getNumResults() < 3)
+        return failure();
+      if (operand.getOperandNumber() == 1) {
+        assert(map.getNumResults() == 5);
+        maps.push_back(map.getMinorSubMap(4));
+      } else {
+        maps.push_back(map.getMinorSubMap(3));
+      }
+    } else {
+      assert(isOutputOperand(linalgOp, operand));
+      if (map.getNumResults() < 2)
+        return failure();
+      maps.push_back(map.getMinorSubMap(2));
+    }
+  }
+  SmallVector<AffineMap> compressedDimMaps = compressUnusedDims(maps);
+  using MapList = ArrayRef<ArrayRef<AffineExpr>>;
+  auto infer = [](MapList m) { return AffineMap::inferFromExprList(m); };
+  AffineExpr r1, p3, p4, r2, r3;
```

Comment:
Not sure I understand this part. Are you looking for two parallel loops and 3 reductions? Why is called p3 and p4?

Reply:
There is an extra trailing reduction dimension in linalg.generic in case of VNNI matmul, so I am checking for it in the maps. I thought the name of expressions doesn't matter, does it? Should I use p4, p5?

Reply:
I've updated the names of affine expressions to p4, p5 instead of p3, p4.